### PR TITLE
feat(integration-templates): improve sharepoint templates

### PR DESCRIPTION
**Pull request by hassan254-prog**
## Describe your changes

- Added mapper functions to both templates.
- Removed the logic of saving `deltaTokens` for incremental sync.
- Since we no longer need the paginate helper function, I adopted the built-in `nango.paginate` function to replace it.

## Issue ticket number and link
[EXT-95](https://linear.app/nango/issue/EXT-95/add-sharepoint-template)
## Checklist before requesting a review (skip if just adding/editing APIs & templates)
- [ ] I added tests, otherwise the reason is: 
- [ ] I added observability, otherwise the reason is:
- [ ] I added analytics, otherwise the reason is:

### CODE REVIEW

File: integration-templates/sharepoint-online/actions/list-sites.ts
Comment (TBonnin) [inline integration-templates/sharepoint-online/actions/list-sites.ts]: I would remove the input all together if it is not being used

File: integration-templates/sharepoint-online/actions/list-sites.ts
Comment (TBonnin) [inline integration-templates/sharepoint-online/actions/list-sites.ts:15]: why is paginate not used anymore?

File: integration-templates/sharepoint-online/mappers/to-file.ts
Comment (TBonnin) [inline integration-templates/sharepoint-online/mappers/to-file.ts]: can it simply be `file.file?.mimeType ?? null`?

File: integration-templates/sharepoint-online/mappers/to-file.ts
Comment (TBonnin) [inline integration-templates/sharepoint-online/mappers/to-file.ts]: same question here.

File: integration-templates/sharepoint-online/mappers/to-site.ts
Comment (TBonnin) [inline integration-templates/sharepoint-online/mappers/to-site.ts:17]: looks like `Site` and `SharePointSite` have the same shape. Is this mapping necessary then?

File: integration-templates/sharepoint-online/nango.yaml
Comment (TBonnin) [inline integration-templates/sharepoint-online/nango.yaml]: looks like the action is simply fetching data. Why moving to a POST method?

File: integration-templates/sharepoint-online/syncs/file-sync.ts
Comment (TBonnin) [inline integration-templates/sharepoint-online/syncs/file-sync.ts]: you don't have to create an intermediary variable by chaining filter and map. ex:
```
siteIdToLists[siteId] = lists
    .filter((list: any) => list.list.template === 'documentLibrary')
    .map((l: any) => l.id)
```
`flatMap` is also an option

File: integration-templates/sharepoint-online/syncs/file-sync.ts
Comment (TBonnin) [inline integration-templates/sharepoint-online/syncs/file-sync.ts]: saving one file at a time might not be the most effective. can we do something like:
```
for await (const listItems of nango.paginate(config)) {
        const allMetadata = Promise.all(listItems.map((item) => fetchDriveItemDetails(nango, siteId, listId, item.id)));
        await nango.batchSave<FileMetadata>([allMetadata], 'FileMetadata');
}
```

File: packages/shared/flows.yaml
Comment (TBonnin) [inline packages/shared/flows.yaml]: same question about POST vs GET

File: integration-templates/sharepoint-online/actions/list-sites.ts
Comment (hassan254-prog) [reply integration-templates/sharepoint-online/actions/list-sites.ts]: Updated this in the [commit](https://github.com/NangoHQ/nango/pull/2495/commits/6385422e1a163b9a87bc615308f11fe8910880b4) 👍

File: integration-templates/sharepoint-online/mappers/to-file.ts
Comment (hassan254-prog) [reply integration-templates/sharepoint-online/mappers/to-file.ts]: Updated this in the [commit](https://github.com/NangoHQ/nango/pull/2495/commits/6385422e1a163b9a87bc615308f11fe8910880b4) 👍

File: integration-templates/sharepoint-online/mappers/to-file.ts
Comment (hassan254-prog) [reply integration-templates/sharepoint-online/mappers/to-file.ts]: same in here 👍

File: integration-templates/sharepoint-online/mappers/to-site.ts
Comment (hassan254-prog) [reply integration-templates/sharepoint-online/mappers/to-site.ts:17]: The `SharepointSite`  interface includes  all the fields returned by sharepoint endpoint that we need to map out to a slimmer `Site` object with only essential ones. Please let me know if this mapping is still necessary.

File: integration-templates/sharepoint-online/nango.yaml
Comment (hassan254-prog) [reply integration-templates/sharepoint-online/nango.yaml]: Updated this as well 👍

File: integration-templates/sharepoint-online/syncs/file-sync.ts
Comment (hassan254-prog) [reply integration-templates/sharepoint-online/syncs/file-sync.ts]: Thanks for the pointing this out🙏! Updated.

File: integration-templates/sharepoint-online/nango.yaml
Comment (khaliqgant) [inline integration-templates/sharepoint-online/nango.yaml:25]:
```suggestion
                output: FileMetadata
                version: 1.0.0
```

Comment (khaliqgant) [reply integration-templates/sharepoint-online/nango.yaml:25]: We can talk about this versioning piece during our next sync

File: integration-templates/sharepoint-online/syncs/file-sync.ts
Comment (hassan254-prog) [reply integration-templates/sharepoint-online/syncs/file-sync.ts]: This implementation will introduce parallel requests, which can potentially defeat the request retry policy. Additionally, it may cause problems due to Sharepoint’s API throttling. Please let me know if we should proceed with this approach?

File: integration-templates/sharepoint-online/actions/list-sites.ts
Comment (hassan254-prog) [reply integration-templates/sharepoint-online/actions/list-sites.ts:15]: Since we expect the number of sites to be very small, there was no need to paginate here.

File: integration-templates/sharepoint-online/nango.yaml
Comment (hassan254-prog) [reply integration-templates/sharepoint-online/nango.yaml:25]: @khaliqgant, sure, let's discuss the versioning piece during our next sync. I think it might be related to the `version@undefined` error I was encountering earlier during `nango deploy`.

**Review by bodinsamuel — APPROVED** [review]

File: integration-templates/sharepoint-online/syncs/file-sync.ts
Comment (TBonnin) [reply integration-templates/sharepoint-online/syncs/file-sync.ts]: good point about concurrency. we can still fetch the details sequentially and make a single call to batchSave instead of one per item. What do you think?
```
for await (const listItems of nango.paginate(config)) {
        const allMetadata = listItems.map((item) => await fetchDriveItemDetails(nango, siteId, listId, item.id));
        await nango.batchSave<FileMetadata>(allMetadata, 'FileMetadata');
}
```

Comment (hassan254-prog) [reply integration-templates/sharepoint-online/syncs/file-sync.ts]: That is a better alternative. However, my initial focus was also on optimizing memory efficiency. Wouldn't this alternative approach potentially impact memory efficiency, especially when we hold data for too long, such as during retries while fetching drive item? What do you think?

Comment (TBonnin) [reply integration-templates/sharepoint-online/syncs/file-sync.ts]: is memory a problem? also how big are pages? If they are a few hundred items I think it is fine to hold them in memory before to batchSave them. The overhead to batchSave (which is a round-trip to persist over the network) would make saving much slower.

**Comment by bodinsamuel** [comment]
Is this ready to merge?

File: integration-templates/sharepoint-online/syncs/file-sync.ts
Comment (hassan254-prog) [reply integration-templates/sharepoint-online/syncs/file-sync.ts]: Yes, in my opinion, we need to consider memory efficiency when writing syncs. We should batch save as soon as possible. However, in this case, we can hold in memory since we're dealing with a slimmer file object and not saving the actual file itself. The response is paginated, so we will always have fewer records to batch save. I've also made a small tweak to the code: we cannot use `await` directly inside a synchronous callback function like `map`. Please take a look at the [changes](https://github.com/NangoHQ/nango/pull/2495/commits/3a0c1ce17b46a86e78475e0c3b2a0279c16e73f8) and provide me with your feedback. Thank you!

**Comment by hassan254-prog** [comment]
@bodinsamuel, waiting for approval from @TBonnin.

**Review by TBonnin — APPROVED** [review]

**Comment by hassan254-prog** [comment]
@bodinsamuel, this looks good to merge.

**Pull request merged**
